File: examples/ios_shell_host/CrosswakeShell/CrosswakeShell.swift
import Foundation
import Combine
import SwiftUI

@MainActor
public final class CrosswakeShell: ObservableObject {
    @Published public var presentation: ShellPresentation = .booting
    public let coordinator: ActivationCoordinator
    private let config: CrosswakeShellConfig
    private var cancellables = Set<AnyCancellable>()

    public init(config: CrosswakeShellConfig, bundle: Bundle = .main) {
        self.config = config
        self.coordinator = ActivationCoordinator.bundled(bundle: bundle, config: config)

        self.coordinator.$presentation
            .receive(on: DispatchQueue.main)
            .sink { [weak self] newPresentation in
                self?.presentation = newPresentation
            }
            .store(in: &cancellables)
    }

    public func createBridgeChannel(
        session: LiveViewSession,
        transferCoordinator: TransferCoordinator?,
        replySink: @escaping (BridgeReplyEnvelope) -> Void
    ) -> BridgeChannel {
        BridgeChannel(
            session: session,
            transferCoordinator: transferCoordinator,
            replySink: replySink,
            config: config
        )
    }

    public func bootstrap(intent: URL? = nil) {
        if let intent = intent {
            coordinator.openURL(intent)
        } else {
            coordinator.bootstrapIfNeeded()
        }
    }
}

File: examples/android_shell_host/app/src/main/java/dev/crosswake/shell/CrosswakeShell.kt
package dev.crosswake.shell

import android.content.Context
import android.content.Intent
import dev.crosswake.shell.transfer.TransferCoordinator
import kotlinx.coroutines.flow.StateFlow

class CrosswakeShell(
    context: Context,
    val config: CrosswakeShellConfig
) {
    private val coordinator = ActivationCoordinator.bundled(context, config)

    val state: StateFlow<ShellPresentation>
        get() = coordinator.stateFlow

    fun bootstrap(intent: Intent? = null): ShellPresentation {
        return coordinator.bootstrapIfNeeded(intent)
    }

    fun handleIntent(intent: Intent?): ShellPresentation {
        return coordinator.handleIntent(intent)
    }
    
    fun retry(): ShellPresentation {
        return coordinator.retry()
    }
    
    fun resolveNavigation(url: String): NavigationDecision {
        return coordinator.resolveNavigation(url)
    }
    
    val currentTransferCoordinator: TransferCoordinator?
        get() = coordinator.currentTransferCoordinator

    fun createBridgeChannel(
        session: LiveViewSession,
        transferCoordinator: TransferCoordinator?
    ): BridgeChannel {
        return BridgeChannel(session, transferCoordinator, config)
    }
}
